packages/core/src/pubsub/pubsub.ts

```diff
@@ -66,6 +69,14 @@ export class Pubsub extends Observable<PubsubMessage> {
     // Textually, `this.peerId$` appears after it is called.
     // Really, subscription is lazy, so `this.peerId$` is populated before the actual subscription act.
     this.peerId$ = from<Promise<string>>(this.ipfs.id().then((_) => _.id))
+    this.pubsubKeepaliveInterval = setInterval(
+      this.publishPubsubKeepaliveIfNeeded.bind(this),
+      this.maxPubsubPublishInterval / 2
+    )
+  }
+
+  shutdown(): void {
```

Comment:
`Pubsub` is `Observable`, do not stuff it with an alien responsibility. See how `MessageBus` treats a similar case.

This retry feature should be a thing between MessageBus and Pubsub, similar to MessageBus in how it treats "I am subscribed to" state.

Reply:
so I have to admit I don't 100% understand what MessageBus is doing with `betweenTimeout`, but from how it's used in `queryNetwork`, I think it's for doing the timeout so that we keep waiting an extra 300ms after the first response is received, to make sure we don't get a better response.  That seems fairly different than what we're trying to do here which is do something on an interval, but skip doing the thing if we've done it already within some interval.

To be honest, I find much of the more complex rxjs logic pretty inscrutable.  The advantage of doing it the way I did it here is it's actually understandable to new readers, since the logic is pretty simple.  If you really think this should be done with rxjs operations, then I might need to hand this ticket off to you, or maybe we can pair program it, because I can't imagine myself being able to figure out how to do it on my own without taking days to do so.